internal/benchrunner/benchrunner.go

```diff
@@ -0,0 +1,204 @@
+// you may not use this file except in compliance with the Elastic License.
+
+package benchrunner
+
+import (
+	"fmt"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/elastic-package/internal/elasticsearch"
+	"github.com/elastic/elastic-package/internal/testrunner"
+)
+
+// BenchType represents the various supported benchmark types
+type BenchType string
+
+// BenchOptions contains benchmark runner options.
+type BenchOptions struct {
+	Folder          testrunner.TestFolder
+	PackageRootPath string
+	API             *elasticsearch.API
+}
+
+// BenchRunner is the interface all benchmark runners must implement.
+type BenchRunner interface {
+	// Type returns the benchmark runner's type.
+	Type() BenchType
+
+	// String returns the human-friendly name of the benchmark runner.
+	String() string
+
+	// Run executes the benchmark runner.
+	Run(BenchOptions) (*Result, error)
+
+	// TearDown cleans up any benchmark runner resources. It must be called
+	// after the benchmark runner has finished executing.
+	TearDown() error
+}
+
+var runners = map[BenchType]BenchRunner{}
+
+// Result contains a single benchmark's results
+type Result struct {
+	// Package to which this benchmark result belongs.
+	Package string
+
+	// BenchType indicates the type of benchmark.
+	BenchType BenchType
+
+	// Data stream to which this benchmark result belongs.
+	DataStream string
+
+	// Time elapsed from running a benchmark case to arriving at its result.
+	TimeElapsed time.Duration
+
+	// If there was an error while running the benchmark case, description
+	// of the error. An error is when the benchmark cannot complete execution due
+	// to an unexpected runtime error in the benchmark execution.
+	ErrorMsg string
+
+	// Benchmark results.
+	Benchmark *BenchmarkResult
+}
+
+// ResultComposer wraps a Result and provides convenience methods for
+// manipulating this Result.
+type ResultComposer struct {
+	Result
+	StartTime time.Time
+}
+
+// NewResultComposer returns a new ResultComposer with the StartTime
+// initialized to now.
+func NewResultComposer(tr Result) *ResultComposer {
+	return &ResultComposer{
+		Result:    tr,
+		StartTime: time.Now(),
+	}
+}
+
+// WithError sets an error on the benchmark result wrapped by ResultComposer.
+func (rc *ResultComposer) WithError(err error) ([]Result, error) {
+	rc.TimeElapsed = time.Since(rc.StartTime)
+	if err == nil {
+		return []Result{rc.Result}, nil
+	}
+
+	rc.ErrorMsg += err.Error()
+	return []Result{rc.Result}, err
+}
+
+// WithSuccess marks the benchmark result wrapped by ResultComposer as successful.
+func (rc *ResultComposer) WithSuccess() ([]Result, error) {
+	return rc.WithError(nil)
+}
+
+// FindBenchmarkFolders finds benchmark folders for the given package and, optionally, benchmark type and data streams
+func FindBenchmarkFolders(packageRootPath string, dataStreams []string, benchType BenchType) ([]testrunner.TestFolder, error) {
```

Comment:
I see that a lot of the benchrunner logic is similar to the testrunner. Frankly speaking, we consider that part of elastic-package as a candidate to be refactored :) I'm wondering if you can extract some common elements of both runners.

If it's a stupid idea, feel free to decline. I know that the logic is tangled here.

Reply:
The reason I preferred to duplicate it in this case is because I am not sure of how this feature might evolve in the near future, so is possible that the logic starts diverging more as time goes by. I am happy to do that if you prefer so now, but given the logic is quite coupled I think it was safer to just assume some level of duplication given both features are isolated and are not intended to evolve together at this point.